File: algorithms/2_medium/C/00958_check_completeness_of_a_binary_tree/bfs.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    static bool isCompleteTree(TreeNode* root) {
        auto q = deque<TreeNode*>();
        q.push_back(root);
        bool seen_null = false;

        while (!q.empty()) {
            auto node = q.front();
            q.pop_front();

            if (node == nullptr) {
                seen_null = true;
            } else {
                if (seen_null) {
                    return false;
                }
                q.push_back(node->left);
                q.push_back(node->right);
            }
        }

        return true;
    }
};

File: algorithms/2_medium/C/00958_check_completeness_of_a_binary_tree/bfs.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::VecDeque;

impl Solution {
    pub fn is_complete_tree(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        if root.is_none() {
            return true;
        }

        let mut q = VecDeque::new();
        q.push_back(root.unwrap());
        let mut seen_last_child = false;

        while let Some(node_rc) = q.pop_front() {
            let node = node_rc.borrow();

            if let Some(left) = node.left.as_ref() {
                if seen_last_child {
                    return false;
                }
                q.push_back(left.clone());
            } else {
                seen_last_child = true;
            }

            if let Some(right) = node.right.as_ref() {
                if seen_last_child {
                    return false;
                }
                q.push_back(right.clone());
            } else {
                seen_last_child = true;
            }
        }

        true
    }
}
